File: app/src/main/java/com/doubean/ford/data/db/GroupDao.java
package com.doubean.ford.data.db;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.RewriteQueriesToDropUnusedColumns;
import androidx.room.Update;
import androidx.room.Upsert;

import com.doubean.ford.data.vo.Group;
import com.doubean.ford.data.vo.GroupBrief;
import com.doubean.ford.data.vo.GroupDetail;
import com.doubean.ford.data.vo.GroupItem;
import com.doubean.ford.data.vo.GroupPostsResult;
import com.doubean.ford.data.vo.GroupSearchResult;
import com.doubean.ford.data.vo.GroupTagPostsResult;
import com.doubean.ford.data.vo.Post;
import com.doubean.ford.data.vo.PostComment;
import com.doubean.ford.data.vo.PostCommentsResult;
import com.doubean.ford.data.vo.PostItem;
import com.doubean.ford.data.vo.PostSortBy;
import com.doubean.ford.data.vo.PostTopComments;

import java.util.Comparator;
import java.util.List;

/**
 * The Data Access Object for the [Group] class.
 * TODO: create base dao?
 */
@Dao
public interface GroupDao {
    @Query(value = "SELECT * FROM groups WHERE id = :groupId")
    @RewriteQueriesToDropUnusedColumns
    LiveData<GroupDetail> loadDetail(String groupId);

    @Upsert(entity = Group.class)
    @RewriteQueriesToDropUnusedColumns
    void upsertDetail(GroupDetail group);

    @Update(entity = Group.class)
    @RewriteQueriesToDropUnusedColumns
    void upsertGroupBrief(GroupBrief group);

    @Query("SELECT * FROM groups")
    @RewriteQueriesToDropUnusedColumns
    LiveData<List<GroupItem>> getGroups();

    @Upsert(entity = Group.class)
    @RewriteQueriesToDropUnusedColumns
    void upsertGroups(List<GroupItem> groupList);

    @Query("SELECT * FROM GroupSearchResult WHERE `query` = :query")
    LiveData<GroupSearchResult> search(String query);

    @Query("SELECT * FROM GroupSearchResult WHERE `query` = :query")
    GroupSearchResult findSearchResult(String query);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertGroupSearchResult(GroupSearchResult groupSearchResult);

    @Query("SELECT * FROM groups WHERE id IN (:groupIds)")
    @RewriteQueriesToDropUnusedColumns
    LiveData<List<GroupItem>> loadGroupsById(List<String> groupIds);

    default LiveData<List<GroupItem>> loadOrderedGroups(List<String> groupIds, Comparator<GroupItem> c) {
        return Transformations.map(loadGroupsById(groupIds), groups -> {
            if (c == null) {
                groups.sort(Comparator.comparingInt(o -> groupIds.indexOf(o.id)));
            } else groups.sort(c);
            return groups;
        });
    }

    default LiveData<List<GroupItem>> loadOrderedGroups(List<String> groupIds) {
        return loadOrderedGroups(groupIds, null);
    }

    /* Posts-related operations start */

    @Query("SELECT * FROM Post WHERE groupId = :groupId ORDER BY created DESC")
    @RewriteQueriesToDropUnusedColumns
    LiveData<List<PostItem>> loadGroupPosts(String groupId);

    @Query("SELECT * FROM Post WHERE groupId = :groupId AND tagId=:tagId ORDER BY created DESC")
    @RewriteQueriesToDropUnusedColumns
    LiveData<List<PostItem>> loadGroupTagPosts(String groupId, String tagId);

    @Query("SELECT * FROM Post WHERE id IN (:postIds)")
    @RewriteQueriesToDropUnusedColumns
    LiveData<List<PostItem>> loadPostsById(List<String> postIds);

    default LiveData<List<PostItem>> loadPosts(List<String> postIds) {
        return Transformations.map(loadPostsById(postIds), posts -> {
            posts.sort(Comparator.comparingInt(o -> postIds.indexOf(o.id)));
            return posts;
        });
    }

    @Query("SELECT * FROM GroupPostsResult WHERE groupId = :groupId AND sortBy = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    LiveData<GroupPostsResult> getGroupPosts(String groupId, PostSortBy sortBy);


    @Query("SELECT * FROM GroupPostsResult WHERE groupId = :groupId AND sortBy = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    GroupPostsResult findGroupPosts(String groupId, PostSortBy sortBy);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertGroupPostsResult(GroupPostsResult groupPostsResult);


    @Query("SELECT * FROM GroupTagPostsResult WHERE groupId = :groupId AND tagId=:tagId AND sortBy = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    LiveData<GroupTagPostsResult> getGroupTagPosts(String groupId, String tagId, PostSortBy sortBy);

    @Query("SELECT * FROM GroupTagPostsResult WHERE groupId = :groupId AND tagId=:tagId AND sortBy = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    GroupTagPostsResult findGroupTagPosts(String groupId, String tagId, PostSortBy sortBy);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertGroupTagPostsResult(GroupTagPostsResult groupTagPostsResult);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertPost(Post post);

    @Upsert(entity = Post.class)
    @RewriteQueriesToDropUnusedColumns
    void upsertPosts(List<PostItem> List);

    @Query("SELECT * FROM Post WHERE id=:postId")
    LiveData<Post> loadPost(String postId);


    @Query("SELECT * FROM post_top_comments WHERE postId = :postId")
    LiveData<PostTopComments> loadPostTopComments(String postId);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertPostTopComments(PostTopComments topComments);

    default LiveData<List<PostComment>> loadOrderedComments(List<String> commentIds, Comparator<PostComment> c) {
        return Transformations.map(loadCommentsById(commentIds), comments -> {
            if (c != null)
                comments.sort(c);
            else
                comments.sort(Comparator.comparingInt(o -> commentIds.indexOf(o.id)));
            return comments;
        });
    }

    default LiveData<List<PostComment>> loadOrderedComments(List<String> commentIds) {
        return loadOrderedComments(commentIds, null);
    }

    @Query("SELECT * FROM post_comments WHERE id IN (:commentIds)")
    LiveData<List<PostComment>> loadCommentsById(List<String> commentIds);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertPostComments(List<PostComment> comments);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertPostCommentsResult(PostCommentsResult postCommentsResult);

    @Query("SELECT * FROM PostCommentsResult WHERE postId=:postId")
    LiveData<PostCommentsResult> getPostComments(String postId);

    @Query("SELECT * FROM PostCommentsResult WHERE postId=:postId")
    PostCommentsResult findPostComments(String postId);
}

File: app/src/main/java/com/doubean/ford/data/db/GroupDao.kt
package com.doubean.ford.data.db

import androidx.room.*
import com.doubean.ford.data.db.model.*
import com.doubean.ford.model.GroupRecommendationType
import com.doubean.ford.model.PostSortBy
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/**
 * The Data Access Object for Group related operations.
 */
@Dao
interface GroupDao {
    @Transaction
    @Query(value = "SELECT * FROM groups WHERE id = :groupId")
    @RewriteQueriesToDropUnusedColumns
    fun loadGroupDetail(groupId: String): Flow<PopulatedGroupDetail>

    @Transaction
    @Query(value = "SELECT * FROM groups WHERE id = :groupId")
    @RewriteQueriesToDropUnusedColumns
    fun getGroupDetail(groupId: String): PopulatedGroupDetail?

    @Upsert(entity = GroupEntity::class)
    @RewriteQueriesToDropUnusedColumns
    suspend fun upsertGroupDetail(group: GroupDetailPartialEntity)

    @Update(entity = GroupEntity::class)
    @RewriteQueriesToDropUnusedColumns
    suspend fun upsertPostGroup(group: PostGroupPartialEntity)

    @Upsert(entity = GroupEntity::class)
    @RewriteQueriesToDropUnusedColumns
    suspend fun upsertSearchResultGroups(groupList: List<GroupSearchResultGroupItemPartialEntity>)

    @Upsert(entity = GroupEntity::class)
    @RewriteQueriesToDropUnusedColumns
    suspend fun upsertRecommendedGroupItemGroups(groupList: List<RecommendedGroupItemGroupPartialEntity>)

    @Query("SELECT * FROM group_search_results WHERE `query` = :query")
    fun loadSearchResult(query: String): Flow<GroupSearchResult?>

    @Query("SELECT * FROM group_search_results WHERE `query` = :query")
    suspend fun getSearchResult(query: String): GroupSearchResult?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertGroupSearchResult(groupSearchResult: GroupSearchResult?)

    @Query("SELECT * FROM groups WHERE id IN (:groupIds)")
    @RewriteQueriesToDropUnusedColumns
    fun loadSearchResultGroups(groupIds: List<String>): Flow<List<GroupSearchResultGroupItemPartialEntity>>

    fun loadOrderedSearchResultGroups(
        groupIds: List<String>,
        c: Comparator<GroupSearchResultGroupItemPartialEntity>? = null,
    ) = this.loadSearchResultGroups(groupIds).map { groups ->
        if (c == null) {
            groups.sortedWith(compareBy { o -> groupIds.indexOf(o.id) })
        } else groups.sortedWith(c)
        //groups
    }

    @Transaction
    @Query("SELECT * FROM posts WHERE id IN (:postIds)")
    @RewriteQueriesToDropUnusedColumns
    fun loadPosts(postIds: List<String>): Flow<List<PopulatedPostItem>>
    fun loadOrderedPosts(postIds: List<String>) = loadPosts(postIds).map { posts ->
        posts.sortedWith(compareBy { o -> postIds.indexOf(o.partialEntity.id) })
    }

    @Transaction
    @Query("SELECT * FROM posts WHERE id IN (:postIds)")
    @RewriteQueriesToDropUnusedColumns
    fun loadPostsWithGroups(postIds: List<String>): Flow<List<PopulatedPostItemWithGroup>>

    fun loadOrderedPostsWithGroups(postIds: List<String>) =
        loadPostsWithGroups(postIds).map { posts ->
            posts.sortedWith(compareBy { o -> postIds.indexOf(o.partialEntity.id) })
        }

    @Query("SELECT * FROM group_posts_result WHERE group_id = :groupId AND sort_by = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    fun loadGroupPosts(groupId: String, sortBy: PostSortBy): Flow<GroupPostsResult?>

    @Query("SELECT * FROM group_posts_result WHERE group_id = :groupId AND sort_by = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    suspend fun getGroupPosts(groupId: String, sortBy: PostSortBy): GroupPostsResult?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertGroupPostsResult(groupPostsResult: GroupPostsResult?)

    @Query("SELECT * FROM group_tag_posts_results WHERE group_id = :groupId AND tag_id=:tagId AND sort_by = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    fun loadGroupTagPosts(
        groupId: String,
        tagId: String,
        sortBy: PostSortBy,
    ): Flow<GroupTagPostsResult?>

    @Query("SELECT * FROM group_tag_posts_results WHERE group_id = :groupId AND tag_id=:tagId AND sort_by = :sortBy")
    @RewriteQueriesToDropUnusedColumns
    suspend fun getGroupTagPosts(
        groupId: String,
        tagId: String,
        sortBy: PostSortBy,
    ): GroupTagPostsResult?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertGroupTagPostsResult(groupTagPostsResult: GroupTagPostsResult?)

    @Insert(entity = PostEntity::class, onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPostDetail(post: PostDetailPartialEntity)

    @Upsert(entity = PostEntity::class)
    @RewriteQueriesToDropUnusedColumns
    suspend fun upsertPosts(posts: List<PostItemPartialEntity>)

    @Transaction
    @Query("SELECT * FROM posts WHERE id=:postId")
    fun loadPost(postId: String): Flow<PopulatedPostDetail>

    fun loadOrderedComments(
        commentIds: List<String>,
        c: Comparator<PopulatedPostComment>? = null,
    ): Flow<List<PopulatedPostComment>> {
        return loadComments(commentIds).map { comments ->
            if (c != null) comments.sortedWith(c) else comments.sortedWith(
                compareBy { o -> commentIds.indexOf(o.entity.id) })
            //comments
        }
    }

    @Transaction
    @Query("SELECT * FROM post_comments WHERE id IN (:commentIds)")
    fun loadComments(commentIds: List<String>): Flow<List<PopulatedPostComment>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPostComments(comments: List<PostCommentEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPostCommentsResult(postCommentsResult: PostCommentsResult?)

    @Query("SELECT * FROM post_comments_results WHERE postId=:postId")
    fun loadPostComments(postId: String): Flow<PostCommentsResult?>

    @Query("SELECT * FROM post_comments_results WHERE postId=:postId")
    fun findPostComments(postId: String): PostCommentsResult?

    @Query("SELECT * FROM recommended_groups_results WHERE recommendation_type=:type")
    fun loadRecommendedGroups(type: GroupRecommendationType): Flow<RecommendedGroupsResult?>

    @Upsert(entity = RecommendedGroupEntity::class)
    suspend fun upsertRecommendedGroups(recommendedGroupEntities: List<RecommendedGroupEntity>)

    @Transaction
    @Query("SELECT * FROM recommended_groups WHERE group_id IN (:ids)")
    @RewriteQueriesToDropUnusedColumns
    fun loadRecommendedGroups(ids: List<String>): Flow<List<PopulatedRecommendedGroup>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertRecommendedGroupsResult(recommendedGroupsResult: RecommendedGroupsResult?)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertGroupTabs(tabs: List<GroupTabEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertGroupPostTags(postTags: List<GroupPostTagEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertRecommendedGroupPosts(recommendedGroupPosts: List<RecommendedGroupPost>)

    @Query("DELETE FROM recommended_group_posts WHERE group_id IN (:groupIds)")
    suspend fun deleteRecommendedGroupPostByGroupIds(groupIds: List<String>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPostTagCrossRefs(postTagCrossRefs: List<PostTagCrossRef>)

    @Query("DELETE FROM posts_tags WHERE post_id =:postId")
    suspend fun deletePostTagCrossRefsByPostId(postId: String)

    @Query("DELETE FROM posts_tags WHERE post_id IN (:postIds)")
    suspend fun deletePostTagCrossRefsByPostIds(postIds: List<String>)
}